Services/Authentication/classes/class.ilSession.php

```diff
@@ -143,54 +140,68 @@ public static function _writeData(string $a_session_id, string $a_data): bool
-            );
-        if ($ilClientIniFile->readVariable("session", "save_ip")) {
-            $fields["remote_addr"] = array("text", $_SERVER["REMOTE_ADDR"]);
+        $fields = [
+            'user_id' => [ilDBConstants::T_INTEGER, (int) (self::get('_authsession_user_id') ?? 0)],
+            'expires' => [ilDBConstants::T_INTEGER, self::getExpireValue()],
+            'data' => [ilDBConstants::T_CLOB, $a_data],
+            'ctime' => [ilDBConstants::T_INTEGER, $now],
+            'type' => [ilDBConstants::T_INTEGER, (int) (self::get('SessionType') ?? 0)]
+        ];
+        if ($ilClientIniFile->readVariable('session', 'save_ip')) {
+            $fields['remote_addr'] = [ilDBConstants::T_TEXT, $_SERVER['REMOTE_ADDR'] ?? ''];
         }
 
         if (self::_exists($a_session_id)) {
             // note that we do this only when inserting the new record
             // updating may get us other contexts for the same session, especially ilContextWAC, which we do not want
-            if (class_exists("ilContext") && ilContext::isSessionMainContext()) {
-                $fields["context"] = array("text", ilContext::getType());
+            if (class_exists('ilContext') && ilContext::isSessionMainContext()) {
+                $fields['context'] = [ilDBConstants::T_TEXT, ilContext::getType()];
             }
             $ilDB->update(
-                "usr_session",
+                'usr_session',
                 $fields,
-                array("session_id" => array("text", $a_session_id))
+                ['session_id' => [ilDBConstants::T_TEXT, $a_session_id]]
             );
         } else {
-            $fields["session_id"] = array("text", $a_session_id);
-            $fields["createtime"] = array("integer", $now);
+            $fields['session_id'] = [ilDBConstants::T_TEXT, $a_session_id];
+            $fields['createtime'] = [ilDBConstants::T_INTEGER, $now];
 
             // note that we do this only when inserting the new record
             // updating may get us other contexts for the same session, especially ilContextWAC, which we do not want
-            if (class_exists("ilContext")) {
-                $fields["context"] = array("text", ilContext::getType());
+            if (class_exists('ilContext')) {
+                $fields['context'] = [ilDBConstants::T_TEXT, ilContext::getType()];
             }
 
-            $ilDB->insert("usr_session", $fields);
+            try {
+                $ilDB->insert('usr_session', $fields);
+            } catch (PDOException $e) {
+                if ($e->getCode() !== self::DUPLICATE_SESSION_DB_ERROR_CODE) {
+                    throw $e;
+                }
+
+                $ilDB->replace(
+                    'usr_session',
+                    [
+                        'session_id' => [ilDBConstants::T_TEXT, $a_session_id]
+                    ],
+                    $fields
+                );
+            }
```

Comment:
Instead of trying to work around an exception in user code, I would prefer to use $ilDB->manipulate with an insert ... on duplicate key clause. Otherwise we could look at extending the $ilDB->insert to facilitate this case.